File: sdk/symbol_dart_sdk/bin/symbol/VotingKeysGenerator.dart
import 'dart:typed_data';
import './KeyPair.dart';
import '../CryptoTypes.dart';

void setBuffer(ByteData destination, int offset, List<int> source) {
  for (var i = 0; i < source.length; i++) {
    destination.setUint8(offset + i, source[i]);
  }
}

class VotingKeysGenerator {
  KeyPair rootKeyPair;
  Function privateKeyGenerator;

  VotingKeysGenerator(KeyPair rootKeyPair, [Function privateKeyGenerator = PrivateKey.random]): 
    privateKeyGenerator = privateKeyGenerator,
    rootKeyPair = rootKeyPair;

  Uint8List generate(int startEpoch, int endEpoch) {
    const HEADER_SIZE = 80;
    const EPOCH_ENTRY_SIZE = 96;

    var numEpochs = endEpoch - startEpoch + 1;
    var buffer = ByteData(HEADER_SIZE + (EPOCH_ENTRY_SIZE * numEpochs));

    buffer.setUint64(0, startEpoch, Endian.little); // start key identifier
    buffer.setUint64(8, endEpoch, Endian.little); // end key identifier
    buffer.setUint64(16, 0xFFFFFFFFFFFFFFFF, Endian.little); // reserved - last (used) key identifier
    buffer.setUint64(24, 0xFFFFFFFFFFFFFFFF, Endian.little); // reserved - last wiped key identifier

    setBuffer(buffer, 32, rootKeyPair.publicKey.bytes); // root voting public key
    buffer.setUint64(64, startEpoch, Endian.little); // level 1/1 start key identifier
    buffer.setUint64(72, endEpoch, Endian.little); // level 1/1 end key identifier

    for (var i = 0; i < numEpochs; i++) {
      var identifier = endEpoch - i;
      var childPrivateKey = privateKeyGenerator();
      var childKeyPair = KeyPair(childPrivateKey);

      var parentSignedPayloadBuffer = ByteData(40);
      setBuffer(parentSignedPayloadBuffer, 0, childKeyPair.publicKey.bytes);
      parentSignedPayloadBuffer.setUint64(32, identifier, Endian.little);
      var signature = rootKeyPair.sign(parentSignedPayloadBuffer.buffer.asUint8List());

      var startOffset = HEADER_SIZE + (EPOCH_ENTRY_SIZE * i);
      setBuffer(buffer, startOffset, childKeyPair.privateKey.bytes); // child voting private key used to sign votes for an epoch
      setBuffer(buffer, startOffset + PrivateKey.SIZE, signature.bytes); // signature proving derivation of child key pair from root
    }

    return buffer.buffer.asUint8List();
  }
}

File: sdk/csharp/SymbolSdk/Symbol/VotingKeysGenerator.cs
namespace SymbolSdk.Symbol;

public class VotingKeysGenerator
{
    public KeyPair RootKeyPair;
    public Func<PrivateKey> PrivateKeyGenerator;
    public VotingKeysGenerator(KeyPair rootKeyPair, Func<PrivateKey>? privateKeyGenerator = null)
    {
        RootKeyPair = rootKeyPair;
        PrivateKeyGenerator = privateKeyGenerator ?? PrivateKey.Random;
    }

    public byte[] Generate(long startEpoch, long endEpoch)
    {
        const byte HEADER_SIZE = 80;
        const byte EPOCH_ENTRY_SIZE = 96;

        var numEpochs = endEpoch - startEpoch + 1;
        var buffer = new byte[HEADER_SIZE + EPOCH_ENTRY_SIZE * numEpochs];
        BitConverter.GetBytes(startEpoch).CopyTo(buffer, 0);
        BitConverter.GetBytes(endEpoch).CopyTo(buffer, 8);
        BitConverter.GetBytes(0xFFFFFFFFFFFFFFFF).CopyTo(buffer, 16);
        BitConverter.GetBytes(0xFFFFFFFFFFFFFFFF).CopyTo(buffer, 24);
        RootKeyPair.PublicKey.bytes.CopyTo(buffer, 32);
        BitConverter.GetBytes(startEpoch).CopyTo(buffer, 64);
        BitConverter.GetBytes(endEpoch).CopyTo(buffer, 72);
        
        for (int i = 0; i < numEpochs; i++) {
            var identifier = endEpoch - i;
            var childPrivateKey = PrivateKeyGenerator();
            var childKeyPair = new KeyPair(childPrivateKey);

            var parentSignedPayloadBuffer = new byte[40];
            childKeyPair.PublicKey.bytes.CopyTo(parentSignedPayloadBuffer, 0);
            BitConverter.GetBytes(identifier).CopyTo(parentSignedPayloadBuffer, 32);
            var signature = RootKeyPair.Sign(parentSignedPayloadBuffer);
            var startOffset = HEADER_SIZE + EPOCH_ENTRY_SIZE * i;
            childKeyPair.PrivateKey.bytes.CopyTo(buffer, startOffset);
            signature.bytes.CopyTo(buffer, startOffset + PrivateKey.SIZE);
        }
        return buffer;
    }
}